### PR TITLE
Make more use of llvm::zip and enumerate in explorer

### PR DESCRIPTION
**Pull request by jonmeow**
I was initially looking at llvm::seq, but then looking through uses it seemed like enumerate/zip would be best for these, allowing less direct indexing and more range-based looping.

### CODE REVIEW

File: explorer/interpreter/interpreter.cpp [hunk 1]
```diff
@@ -1534,17 +1530,18 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
...
-          int i = 1 + num_args;
-          for (const auto& [impl_bind, impl_exp] : call.witnesses()) {
-            witnesses[impl_bind] = act.results()[i];
-            ++i;
+          for (const auto& [witness, result] : llvm::zip(
+                   call.witnesses(),
+                   llvm::ArrayRef(act.results()).drop_front(1 + num_args))) {
+            witnesses[witness.first] = result;
           }
         }
```
Comment (zygoloid) [inline explorer/interpreter/interpreter.cpp]: Maybe we should consider adding a `Split` utility so we can do this:

```suggestion
        std::vector<Nonnull<const Value*>> param_values;
        auto [arg_results, witness_results] =
            Split(llvm::ArrayRef(act.results()).drop_front(), num_args);
        for (const auto& arg_result : arg_results) {
          param_values.push_back(arg_result);
        }
        const auto* param_tuple = arena_->New<TupleValue>(param_values);
        // Prepare witnesses.
        ImplWitnessMap witnesses;
        if (num_witnesses > 0) {
          for (const auto& [witness, result] : llvm::zip(call.witnesses(), witness_results)) {
            witnesses[witness.first] = result;
          }
        }
```

(No action expected in this PR though.)

File: explorer/interpreter/interpreter.cpp [hunk 2]
```diff
@@ -290,10 +290,9 @@ auto Interpreter::EvalPrim(Operator op, Nonnull<const Value*> /*static_type*/,
 auto Interpreter::CreateStruct(const std::vector<FieldInitializer>& fields,
                                const std::vector<Nonnull<const Value*>>& values)
     -> Nonnull<const Value*> {
-  CARBON_CHECK(fields.size() == values.size());
   std::vector<NamedValue> elements;
-  for (size_t i = 0; i < fields.size(); ++i) {
-    elements.push_back({fields[i].name(), values[i]});
+  for (const auto& [field, value] : llvm::zip_equal(fields, values)) {
```
Comment (zygoloid) [inline explorer/interpreter/interpreter.cpp]: Throughout this patch:

Using `const auto&` here seems unnecessary to me: the iterator will be generating new values, so a non-reference would do the same thing. Maybe remove the `&`s throughout?

**Review by zygoloid — APPROVED** [review]

File: explorer/interpreter/interpreter.cpp (hunk 2 above)
Comment (jonmeow) [reply explorer/interpreter/interpreter.cpp]: Done

File: explorer/interpreter/interpreter.cpp (hunk 1 above)
Comment (jonmeow) [reply explorer/interpreter/interpreter.cpp]: Acknowledged, although it seems not as worth it if we don't get more use-cases... (though, I can use `slice`, might go and fix that)

**Pull request merged**